megatron/model/adapter.py

```diff
@@ -113,9 +140,41 @@ def forward(self, x: TensorType["b", "s", "d"], *attn_args, **attn_kwargs):
             attn_outputs[0],
             attn_outputs[1:],
         )  # outputs: output, bias
-        hidden_states = self.adapter(attn_output) + attn_output
+        hidden_states = self.adapt(attn_output) + attn_output
         return (hidden_states,) + outputs
 
+class ParallelLinearPEFT(torch.nn.Module):
+    
+    def __init__(self, 
+                 par_lin,
+                 rank = 8,
+                 is_lora = False,#Assume it's not a lora. If not, rank = downsampling
+                 alpha = 1.0):
+       
+        super().__init__()
+        self.in_features = par_lin.input_size
+        self.out_features = par_lin.output_size
+        
+        self.par_lin = par_lin
+        self.rank = rank
+        
+        self.downsample_adapter_lora = torch.nn.Parameter(torch.zeros((rank, self.in_features)))
+        self.upsample_adapter_lora = torch.nn.Parameter(torch.zeros((self.out_features, rank)))
+        self.scaling = alpha / self.rank
+        
+        # initialize A the same way as the default for nn.Linear and B to zero
+        #This doesn't matter for domain adaptation
+        #torch.nn.init.kaiming_uniform_(self.downsample_adapter_lora, a=math.sqrt(5))
+        #torch.nn.init.zeros_(self.upsample_adapter_lora)
+    
+    def forward(self, x):
+        
+        result, bias = self.par_lin.forward(x)
+        
+        result += x @ self.downsample_adapter_lora.T @ self.upsample_adapter_lora.T * self.scaling
```

Comment:
Avoid using inplace operations in Torch. Not a big issue here since original result is anyways frozen, but better to avoid this.